src/combinatorics/double_counting.lean

```diff
@@ -0,0 +1,83 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import algebra.big_operators.order
+
+/-!
+# Double countings
+
+This file gathers a few double counting arguments.
+
+## Bipartite graphs
+
+In a bipartite graph (considered as a relation `r : α → β → Prop`), we can bound the number of edges
+between `s : finset α` and `t : finset β` by the minimum/maximum of edges over all `a ∈ s` times the
+the size of `s`. Similarly for `t`. Combining those two yields inequalities between the sizes of `s`
+and `t`.
+
+* `bipartite_below`: `s.bipartite_below r b` are the elements of `s` below `b` wrt to `r`. Its size
+  is the number of edges of `b` in `s`.
+* `bipartite_above`: `t.bipartite_above r a` are the elements of `t` above `a` wrt to `r`. Its size
+  is the number of edges of `a` in `t`.
+* `card_mul_le_card_mul`, `card_mul_le_card_mul'`: Double counting the edges of a bipartite graph
+  from below and from above.
+* `card_mul_eq_card_mul`: Equality combination of the previous.
+-/
+
+open finset function
+open_locale big_operators
+
+/-! ### Bipartite graph -/
+
+namespace finset
+section bipartite
+variables {α β : Type*} (r : α → β → Prop) (s : finset α) (t : finset β) (a a' : α) (b b' : β)
+  [decidable_pred (r a)] [Π a, decidable (r a b)] {m n : ℕ}
+
+/-- Elements of `s` which are "below" `b` according to relation `r`. -/
+def bipartite_below : finset α := s.filter (λ a, r a b)
+
+/-- Elements of `t` which are "above" `a` according to relation `r`. -/
+def bipartite_above : finset β := t.filter (r a)
+
+lemma bipartite_below_swap : t.bipartite_below (swap r) a = t.bipartite_above r a := rfl
+lemma bipartite_above_swap : s.bipartite_above (swap r) b = s.bipartite_below r b := rfl
```

Comment:
Should these be `@[simp]`?

Reply:
No, because they always apply. I too originally got tricked into thinking the LHS is simpler than the RHS. It's actually not.